sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala

```diff
@@ -266,28 +267,41 @@ class ArtifactManager(session: SparkSession) extends AutoCloseable with Logging
    * they are from a permanent location.
    */
   private[sql] def addLocalArtifacts(artifacts: Seq[Artifact]): Unit = {
+    val failedArtifactExceptions = ListBuffer[SparkRuntimeException]()
+
     artifacts.foreach { artifact =>
-      artifact.storage match {
-        case d: Artifact.LocalFile =>
-          addArtifact(
-            artifact.path,
-            d.path,
-            fragment = None,
-            deleteStagedFile = false)
-        case d: Artifact.InMemory =>
-          val tempDir = Utils.createTempDir().toPath
-          val tempFile = tempDir.resolve(artifact.path.getFileName)
-          val outStream = Files.newOutputStream(tempFile)
-          Utils.tryWithSafeFinallyAndFailureCallbacks {
-            d.stream.transferTo(outStream)
-            addArtifact(artifact.path, tempFile, fragment = None)
-          }(finallyBlock = {
-            outStream.close()
-          })
-        case _ =>
-          throw SparkException.internalError(s"Unsupported artifact storage: ${artifact.storage}")
+      try {
+        artifact.storage match {
+          case d: Artifact.LocalFile =>
+            addArtifact(
+              artifact.path,
+              d.path,
+              fragment = None,
+              deleteStagedFile = false)
+          case d: Artifact.InMemory =>
+            val tempDir = Utils.createTempDir().toPath
+            val tempFile = tempDir.resolve(artifact.path.getFileName)
+            val outStream = Files.newOutputStream(tempFile)
+            Utils.tryWithSafeFinallyAndFailureCallbacks {
+              d.stream.transferTo(outStream)
+              addArtifact(artifact.path, tempFile, fragment = None)
+            }(finallyBlock = {
+              outStream.close()
+            })
+          case _ =>
+            throw SparkException.internalError(s"Unsupported artifact storage: ${artifact.storage}")
+        }
+      } catch {
+        case e: SparkRuntimeException =>
+          failedArtifactExceptions += e
       }
     }
+
+    if (failedArtifactExceptions.nonEmpty) {
```

Comment:
The error handling and suppression logic seems to be duplicated in both ArtifactManager.scala and SparkConnectAddArtifactsHandler.scala.
```scala
 if (failedArtifactExceptions.nonEmpty) {
      val exception = failedArtifactExceptions.head
      failedArtifactExceptions.drop(1).foreach(exception.addSuppressed(_))
      throw exception
    }
```
I was wondering whether it makes sense to write a small utility to handle Seq[Try[...]]

Reply:
[ArtifactUtils](https://github.com/apache/spark/blob/f0a3a2ea161181c424d083af76607c6384b2426e/sql/api/src/main/scala/org/apache/spark/sql/util/ArtifactUtils.scala#L4) would be a good place for this utility method